fix: Bank module init genesis optimization

YunSuk-Yeo opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #9425 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

marbar3778 commented on x/bank/keeper/send.go:
```diff
@@ -242,6 +242,25 @@ func (k BaseSendKeeper) clearBalances(ctx sdk.Context, addr sdk.AccAddress) {
+
+	return nil
+}
+
```
Seems `setBalances` is only used in InitGenesis, @amaurym do you think we could remove it outright?

marbar3778 replied:
@YunSuk-Yeo could you remove clearBalances and setBalances then add a changelog entry, please

YunSuk-Yeo replied:
sure updated

marbar3778 commented:
Seems linting is failing, could you fix this, then we can work on merging it.

YunSuk-Yeo commented:
I fixed the lint error in bank module, which is I changed. There are more place the lint error happens.

```
golangci-lint run --out-format=tab
WARN [runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive. 
x/distribution/legacy/v043/helpers.go:22:18  gocritic  appendAssign: append result not assigned to the same slice
make: *** [lint] Error 1
```

alexanderbez commented:
@aaronc we should make sure this gets into the next set of release(s) and backported appropriately.

aaronc commented on x/bank/keeper/send.go:
```diff
@@ -227,31 +227,20 @@ func (k BaseSendKeeper) addCoins(ctx sdk.Context, addr sdk.AccAddress, amt sdk.C
-// clear out all balances prior to setting the new coins as to set existing balances
-// to zero if they don't exist in amt. An error is returned upon failure.
-func (k BaseSendKeeper) setBalances(ctx sdk.Context, addr sdk.AccAddress, balances sdk.Coins) error {
-	k.clearBalances(ctx, addr)
```
Are there any cases where we actually need to clear balances for correctness?
